tests/integration/test_charm.py

```diff
@@ -4,148 +4,158 @@
+    # Insert values into a table on the primary unit
+    random_unit = ops_test.model.applications[APP_NAME].units[0]
+    server_config_credentials = await get_server_config_credentials(random_unit)
+
+    primary_unit = await get_primary_unit(
+        ops_test,
+        random_unit,
+        APP_NAME,
     )
-    assert len(ops_test.model.applications[APP_NAME].units) == 1
-    assert ops_test.model.applications[APP_NAME].units[0].workload_status == "active"
+    primary_unit_address = await get_unit_address(ops_test, primary_unit.name)
 
+    random_chars = generate_random_string(40)
+    create_records_sql = [
+        "CREATE DATABASE IF NOT EXISTS test",
+        "CREATE TABLE IF NOT EXISTS test.data_replication_table (id varchar(40), primary key(id))",
+        f"INSERT INTO test.data_replication_table VALUES ('{random_chars}')",
+    ]
 
-@pytest.mark.skip_if_deployed
-@pytest.mark.abort_on_fail
-async def test_scale_to_ha(ops_test: OpsTest):
-    """Scale the charm to HA."""
-    await ops_test.model.applications[APP_NAME].scale(len(UNIT_IDS))
-
-    # Blocks until all units are in active state
-    logger.info("Wait for app active status")
-    sleep(10)
-    await ops_test.model.wait_for_idle(
-        apps=[APP_NAME],
-        status="active",
-        raise_on_blocked=True,
-        timeout=1000,
+    await execute_queries_on_unit(
+        primary_unit_address,
+        server_config_credentials["username"],
+        server_config_credentials["password"],
+        create_records_sql,
+        commit=True,
     )
 
-    assert len(ops_test.model.applications[APP_NAME].units) == len(UNIT_IDS)
-
-    root_password = await get_password(ops_test, "root-password")
-    # Primary will be ID 0, since the unit is deployed first
-    host_ip = await get_pod_ip(ops_test, f"{APP_NAME}/0")
-    # connect to the MySQL server
-    cnx = db_connect(host_ip, root_password)
-    cursor = cnx.cursor()
-    query = "SELECT count(*) FROM performance_schema.replication_group_members where MEMBER_STATE='ONLINE';"
-    cursor.execute(query)
-
-    result = cursor.fetchone()
-    assert result[0] == len(UNIT_IDS)
-    cursor.close()
-    cnx.close()
-
-
-async def test_replicated_write_n_read(ops_test: OpsTest):
-    """Write on primary and read on secondary."""
-    # get the cluster admin password
-    root_password = await get_password(ops_test, "root-password")
-    # Primary will be ID 0, since the unit is deployed first
-    host_ip = await get_pod_ip(ops_test, f"{APP_NAME}/0")
-    # connect to the MySQL server
-    cnx = db_connect(host_ip, root_password)
-    cursor = cnx.cursor()
-    # ensure table cleanup (when running with --no-deploy)
-    # create a table and insert a row
-    queries = [
-        "DROP TABLE IF EXISTS mysql.charmtest;",
-        "CREATE TABLE mysql.charmtest (test_field VARCHAR(255) PRIMARY KEY);",
-        "INSERT INTO mysql.charmtest VALUES ('hello');",
+    # Allow time for the data to be replicated
+    time.sleep(2)
```

Comment:
Maybe you can change it to something like below to enclose the `execute_queries_on_unit` calls:

```python
try:
    for attempt in Retrying(stop=stop_after_delay(10), wait=wait_fixed(3)):
        with attempt:
            # Execute the logic that can fail until some time has passed.
except RetryError:
    return False
```

Reply:
Added in a9bd5bb1a336943a9c0d601492f594aaddd073e8